### PR TITLE
Ensure WaitUntilFullyLoadedAsync can make progress from a background thread

### PR DESCRIPTION
**Pull request by sharwell**
Fixes [AB#1230148](https://devdiv.visualstudio.com/0bdbc590-a062-4c3f-b0f6-9383f67865ee/_workitems/edit/1230148)

### CODE REVIEW

File: src/VisualStudio/Core/Def/Implementation/Workspace/VisualStudioWorkspaceStatusServiceFactory.cs [hunk 1]
```diff
@@ -140,48 +157,14 @@ public async Task<bool> IsFullyLoadedAsync(CancellationToken cancellationToken)
...
-                var service = await _serviceProvider.GetServiceAsync<SVsOperationProgress, IVsOperationProgressStatusService>(throwOnFailure: false)
-                                                    .WithCancellation(cancellationToken).ConfigureAwait(false);
-
-                return service?.GetStageStatusForSolutionLoad(CommonOperationProgressStageIds.Intellisense);
-            }
-
-            private async Task EnsureInitializationAsync(CancellationToken cancellationToken)
+            private async ValueTask<IVsOperationProgressStageStatusForSolutionLoad?> GetProgressStageStatusAsync(CancellationToken cancellationToken)
             {
-                using (await _initializationGate.DisposableWaitAsync(cancellationToken).ConfigureAwait(false))
```
Comment (sharwell) [inline src/VisualStudio/Core/Def/Implementation/Workspace/VisualStudioWorkspaceStatusServiceFactory.cs:153]: 📝 The problem was this `SemaphoreSlim`. JTF did not know that it needed to let specific work onto the UI thread in order to make progress such that the semaphore is available again, so it blocked the UI thread until the Cancel button was pressed. The new code eliminates this semaphore.

File: src/VisualStudio/Core/Def/Implementation/Workspace/VisualStudioWorkspaceStatusServiceFactory.cs [hunk 2]
```diff
@@ -140,48 +157,14 @@ public async Task<bool> IsFullyLoadedAsync(CancellationToken cancellationToken)
...
+                if (_progressStageStatus.Task.IsCompleted)
```
Comment (sharwell) [inline src/VisualStudio/Core/Def/Implementation/Workspace/VisualStudioWorkspaceStatusServiceFactory.cs:162]: 📝 This branch can be removed if https://github.com/microsoft/vs-threading/pull/696 is approved.

File: src/VisualStudio/Core/Def/Implementation/Workspace/VisualStudioWorkspaceStatusServiceFactory.cs [hunk 3]
```diff
@@ -69,30 +66,54 @@ public IWorkspaceService CreateService(HostWorkspaceServices workspaceServices)
         /// </summary>
         private class Service : IWorkspaceStatusService
         {
-            private readonly SemaphoreSlim _initializationGate = new(initialCount: 1);
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Implementation/Workspace/VisualStudioWorkspaceStatusServiceFactory.cs:72]: nice.

File: src/VisualStudio/Core/Def/Implementation/Workspace/VisualStudioWorkspaceStatusServiceFactory.cs [hunk 4]
```diff
@@ -69,30 +66,54 @@ public IWorkspaceService CreateService(HostWorkspaceServices workspaceServices)
...
-
             /// <summary>
             /// A task indicating that the <see cref="Guids.GlobalHubClientPackageGuid"/> package has loaded. Calls to
             /// <see cref="IServiceBroker.GetProxyAsync"/> may have a main thread dependency if the proffering package
             /// is not loaded prior to the call.
             /// </summary>
-            private JoinableTask _loadHubClientPackage;
+            private readonly JoinableTask _loadHubClientPackage;
+
+            private readonly JoinableTask<IVsOperationProgressStageStatusForSolutionLoad?> _progressStageStatus;
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Implementation/Workspace/VisualStudioWorkspaceStatusServiceFactory.cs:79]: consider docs.

File: src/VisualStudio/Core/Def/Implementation/Workspace/VisualStudioWorkspaceStatusServiceFactory.cs [hunk 5]
```diff
@@ -69,30 +66,54 @@ public IWorkspaceService CreateService(HostWorkspaceServices workspaceServices)
...
                 _threadingContext = threadingContext;
 
-                // pre-emptively make sure event is subscribed. if APIs are called before it is done, calls will be blocked
-                // until event subscription is done
-                var asyncToken = listener.BeginAsyncOperation("StatusChanged_EventSubscription");
-                Task.Run(() => EnsureInitializationAsync(CancellationToken.None), CancellationToken.None).CompletesAsyncOperation(asyncToken);
+                _loadHubClientPackage = _threadingContext.JoinableTaskFactory.RunAsync(async () =>
+                {
+                    // Use the disposal token, since the caller's cancellation token will apply instead to the
+                    // JoinAsync operation.
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Implementation/Workspace/VisualStudioWorkspaceStatusServiceFactory.cs:91]:
```suggestion
                    // JoinAsync operation in GetProgressStageStatusAsync.
```

File: src/VisualStudio/Core/Def/Implementation/Workspace/VisualStudioWorkspaceStatusServiceFactory.cs [hunk 6]
```diff
@@ -69,30 +66,54 @@ public IWorkspaceService CreateService(HostWorkspaceServices workspaceServices)
...
+                    await _threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(alwaysYield: true, _threadingContext.DisposalToken);
+
+                    // Make sure the HubClient package is loaded, since we rely on it for proffered OOP services
+                    var shell = await _serviceProvider.GetServiceAsync<SVsShell, IVsShell7>().ConfigureAwait(true);
+                    Assumes.Present(shell);
+
+                    await shell.LoadPackageAsync(Guids.GlobalHubClientPackageGuid);
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Implementation/Workspace/VisualStudioWorkspaceStatusServiceFactory.cs:98]: i'm not a fan of the .ConfigureAwait inconsistency.  can we either have .ConfigureAwait(true) on both of tehse, or on neither of this?

File: src/VisualStudio/Core/Def/Implementation/Workspace/VisualStudioWorkspaceStatusServiceFactory.cs [hunk 7]
```diff
@@ -140,48 +157,14 @@ public async Task<bool> IsFullyLoadedAsync(CancellationToken cancellationToken)
...
-                    if (status == null)
-                    {
-                        return;
-                    }
-
-                    status.PropertyChanged += (_, e) => this.StatusChanged?.Invoke(this, EventArgs.Empty);
+                    return await _progressStageStatus.Task.ConfigureAwait(false);
                 }
+
+                return await _progressStageStatus.JoinAsync(cancellationToken).ConfigureAwait(false);
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Implementation/Workspace/VisualStudioWorkspaceStatusServiceFactory.cs:167]:
```suggestion
                    return await _progressStageStatus.Task.ConfigureAwait(false);
                }

                return await _progressStageStatus.JoinAsync(cancellationToken).ConfigureAwait(false);
```

i don't actually understand this bit...   why not just always do: 

```return await _progressStageStatus.Task.ConfigureAwait(false);```

? 

If tehre is substantive difference, please explain it with a comment in the code as hte pattern is not self evident to me.  Thanks!

File: src/VisualStudio/Core/Def/Implementation/Workspace/VisualStudioWorkspaceStatusServiceFactory.cs (hunk 6 above)
Comment (sharwell) [reply src/VisualStudio/Core/Def/Implementation/Workspace/VisualStudioWorkspaceStatusServiceFactory.cs:98]: `IVsTask` doesn't have a `ConfigureAwait` extension. Happy to remove the other one once the analyzer is updated to stop analyzing code using JTF thread switching.

File: src/VisualStudio/Core/Def/Implementation/Workspace/VisualStudioWorkspaceStatusServiceFactory.cs (hunk 7 above)
Comment (sharwell) [reply src/VisualStudio/Core/Def/Implementation/Workspace/VisualStudioWorkspaceStatusServiceFactory.cs:167]: > i don't actually understand this bit... why not just always do:
>
> `return await _progressStageStatus.Task.ConfigureAwait(false);`

This is never valid (violation of the third threading rule):

https://github.com/microsoft/vs-threading/blob/master/doc/threading_rules.md#rule-3-if-ever-awaiting-work-that-was-started-earlier-that-work-must-be-joined

It is likely that this optimization path can be removed once vs-threading updates `JoinAsync` to have a proper fast path for completed tasks.

Comment (CyrusNajmabadi) [reply src/VisualStudio/Core/Def/Implementation/Workspace/VisualStudioWorkspaceStatusServiceFactory.cs:167]: i don't actually see how it's a violation of the third rule.  it seems to also say that it's fine to do `await longRunningAsyncWork;`  So i would expect to just see that here.

Comment (sharwell) [reply src/VisualStudio/Core/Def/Implementation/Workspace/VisualStudioWorkspaceStatusServiceFactory.cs:167]: ➡️ Added a comment clarifying that the strange code paths are a workaround for a specific limitation which should be addressed in the future.

File: src/VisualStudio/Core/Def/Implementation/Workspace/VisualStudioWorkspaceStatusServiceFactory.cs (hunk 5 above)
Comment (sharwell) [reply src/VisualStudio/Core/Def/Implementation/Workspace/VisualStudioWorkspaceStatusServiceFactory.cs]: ➡️ Done!

File: src/VisualStudio/Core/Def/Implementation/Workspace/VisualStudioWorkspaceStatusServiceFactory.cs (hunk 4 above)
Comment (sharwell) [reply src/VisualStudio/Core/Def/Implementation/Workspace/VisualStudioWorkspaceStatusServiceFactory.cs:84]: ➡️ Added!

File: src/VisualStudio/Core/Def/Implementation/Workspace/VisualStudioWorkspaceStatusServiceFactory.cs (hunk 7 above)
Comment (sharwell) [reply src/VisualStudio/Core/Def/Implementation/Workspace/VisualStudioWorkspaceStatusServiceFactory.cs:174]: 🔗 Also filed https://github.com/microsoft/vs-threading/issues/697 to make the threading rules more clear about this syntax.

File: src/VisualStudio/Core/Def/Implementation/Workspace/VisualStudioWorkspaceStatusServiceFactory.cs (hunk 6 above)
Comment (AArnott) [reply src/VisualStudio/Core/Def/Implementation/Workspace/VisualStudioWorkspaceStatusServiceFactory.cs:103]: Interesting point. Feel free to file a bug on the devdiv AzDO account to add `.ConfigureAwait()` extension methods to the IVsTask type.

**Pull request merged**